src/data.ml

```diff
@@ -86,14 +86,12 @@ module Entry = struct
     let decode_value string off = V.decode string (off + K.encoded_size)
 
     let encode' key value f =
-      let encoded_key = K.encode key in
-      let encoded_value = V.encode value in
-      if String.length encoded_key <> K.encoded_size then
-        raise (Invalid_size encoded_key);
-      if String.length encoded_value <> V.encoded_size then
-        raise (Invalid_size encoded_value);
-      f encoded_key;
-      f encoded_value
+      K.encode key (fun s ->
+          if String.length s <> K.encoded_size then raise (Invalid_size s));
+      V.encode value (fun s ->
+          if String.length s <> K.encoded_size then raise (Invalid_size s));
+      K.encode key f;
+      V.encode value f
```

Comment:
You're encoding the values twice here; can you make a single call to `{K,V}.encode`?

Reply:
Done ;-)